python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -0,0 +1,428 @@
+        if bw == 32:
+            irt = ir.FloatType()
+        elif bw == 64:
+            irt = ir.DoubleType()
+    return irt
+
+
+@intrinsic
+def load_cs(typingctx, base):
+    # Corresponding to `LOAD_CS` here:
+    # https://nvidia.github.io/cccl/cub/api/classcub_1_1CacheModifiedInputIterator.html
+    def codegen(context, builder, sig, args):
+        rt = _ir_type_given_numba_type(sig.return_type)
+        if rt is None:
+            raise RuntimeError(f"Unsupported: {type(sig.return_type)=}")
+        ftype = ir.FunctionType(rt, [rt.as_pointer()])
+        bw = sig.return_type.bitwidth
+        asm_txt = f"ld.global.cs.b{bw} $0, [$1];"
+        if bw < 64:
+            constraint = "=r, l"
+        else:
+            constraint = "=l, l"
+        asm_ir = ir.InlineAsm(ftype, asm_txt, constraint)
+        return builder.call(asm_ir, args)
+
+    return base.dtype(base), codegen
+
+
+def cache_advance(this, distance):
+    this[0] = this[0] + distance
+
+
+def cache_dereference(this):
+    return load_cs(this[0])
+
+
+class CacheModifiedPointer:
+    def __init__(self, ptr, ntype):
+        self.val = ctypes.c_void_p(ptr)
+        self.ntype = ntype
+        data_as_ntype_pp = numba.types.CPointer(numba.types.CPointer(ntype))
+        self.prefix = "cache" + ntype.name
+        self.ltoirs = [
+            _ncc(
+                f"{self.prefix}_advance",
+                cache_advance,
+                numba.types.void(data_as_ntype_pp, _DISTANCE_NUMBA_TYPE),
+            ),
+            _ncc(
+                f"{self.prefix}_dereference", cache_dereference, ntype(data_as_ntype_pp)
+            ),
+        ]
+
+    @property
+    def state_c_void_p(self):
+        return ctypes.cast(ctypes.pointer(self.val), ctypes.c_void_p)
+
+    @property
+    def size(self):
+        return _DEVICE_POINTER_SIZE
+
+    @property
+    def alignment(self):
+        return _DEVICE_POINTER_SIZE
+
+
+def constant_advance(this, _):
+    pass
+
+
+def constant_dereference(this):
+    return this[0]
+
+
+class ConstantIterator:
+    def __init__(self, val, ntype):
+        thisty = numba.types.CPointer(ntype)
+        self.val = _ctypes_type_given_numba_type(ntype)(val)
+        self.ntype = ntype
+        self.prefix = "constant_" + ntype.name
+        self.ltoirs = [
+            _ncc(
+                f"{self.prefix}_advance",
+                constant_advance,
+                numba.types.void(thisty, _DISTANCE_NUMBA_TYPE),
+            ),
+            _ncc(f"{self.prefix}_dereference", constant_dereference, ntype(thisty)),
+        ]
+
+    @property
+    def state_c_void_p(self):
+        return ctypes.cast(ctypes.pointer(self.val), ctypes.c_void_p)
+
+    @property
+    def size(self):
+        return self.ntype.bitwidth // 8
+
+    @property
+    def alignment(self):
+        return self.size
+
+
+def count_advance(this, diff):
+    this[0] += diff
+
+
+def count_dereference(this):
+    return this[0]
```

Comment:
Done in commit c3c51a56aaa6b4866a4fd08012858d42e347eb5f

I added comments:
```
# Exclusively for numba.cuda.compile (this is not an actual method).
```

My thinking:

Not adding a decorator (as we had originally), people will think it's a bound method, but wonder why `self` is called `this`.

Explicitly adding `@staticmethod` will make people believe it really is a static method, but that's not actually true.

Being explicit in the comment is only slightly more verbose than adding a decorator but much more informative.

Reply:
> Explicitly adding @staticmethod will make people believe it really is a static method, but that's not actually true.

I don't think I understand. If anything, adding `@staticmethod` will make it even _more_ obvious to the reader that the function i independent of the class. Typically functions that have no dependency on the class or its members, but are otherwise related to it are defined as `@staticmethod`.

Reply:
In other words, these are truly `staticmethods` in every sense